### PR TITLE
[ELF] Implement --force-group-allocation

### PR DESCRIPTION
**Pull request by MaskRay**
GNU ld's relocatable linking behaviors:

* Sections with the `SHF_GROUP` flag are handled like sections matched
  by the `--unique=pattern` option. They are processed like orphan
  sections and ignored by input section descriptions.
* Section groups' (usually named `.group`) content is updated as the
  section indexes are updated. Section groups can be discarded with
  `/DISCARD/ : { *(.group) }`.

`-r --force-group-allocation` discards section groups and allows
sections with the `SHF_GROUP` flag to be matched like normal sections.
If two section group members are placed into the same output section,
their relocation sections (if present) are combined as well.
This behavior can be useful when -r output is used as a pseudo shared
object (e.g., FreeBSD's amd64 kernel modules, CHERIoT compartments).

This patch implements --force-group-allocation:

* Input SHT_GROUP sections are discarded.
* Input sections do not get the SHF_GROUP flag, so `addInputSec`
  will combine relocation sections if their relocated section group
  members are combined.

The default behavior is:

* Input SHT_GROUP sections are retained.
* Input SHF_GROUP sections can be matched (unlike GNU ld)
* Input SHF_GROUP sections keep the SHF_GROUP flag, so `addInputSec`
  will create different OutputDesc copies.

GNU ld provides the `FORCE_GROUP_ALLOCATION` command, which is not
implemented.

### CODE REVIEW

File: lld/ELF/Driver.cpp [hunk 1]
```diff
@@ -1384,6 +1384,9 @@ static void readConfigs(opt::InputArgList &args) {
   config->relaxGP = args.hasFlag(OPT_relax_gp, OPT_no_relax_gp, false);
   config->rpath = getRpath(args);
   config->relocatable = args.hasArg(OPT_relocatable);
+  config->resolveGroups = !args.hasArg(OPT_relocatable) ||
```
Comment (smithp35) [inline lld/ELF/Driver.cpp:1387]: I'm not sure I've understood the benefit of `--inhibit-group-allocation`. 

* For a non relocatable link it is ignored. 
* For a relocatable link its behaviour is the default state, so there is no need to explicitly use it.

The only reason I can think to use it if there is an existing linker command-line that with `--force-group-allocation` that I want to disable. I don't think that is likely though.

Unless I've missed something, would it be better to keep the command-line interface simple with 
```
config->resolveGroups = !args.hasArg(OPT_relocatable) ||
args.hasFlag(OPT_force_group_allocation, false);

I couldn't find it in the GNU ld documentation.

**Review by smithp35 — COMMENTED** [review]
Code changes LGTM, I'm not sure we need `--inhibit-group-allocation` although I could have missed a use case.

**Comment by justincady** [comment]
First, thank you so much for doing this. :)

I applied this patch to 17.0.6 to compare behavior. The group sections portion appears to work as expected, but I'm confused by one aspect of the relocation section combining:

```
$ cat b.h
#pragma once
struct A {
    A();
    virtual ~A();
    virtual int foo();
};
struct B : public A {
        int foo() override;
};

$ cat b.cc
#include "b.h"
int A::foo() { return 42; }
int B::foo() { return 84; }

$ cat ldscript.amd64
SECTIONS { .text : { *(.text .stub .text.* .gnu.linkonce.t.*) } }

$ cat rela.sh
#!/bin/bash
clang -c b.cc
ld.lld -r --force-group-allocation -T ldscript.amd64 b.o -o rela.fga.ro

$ ./rela.sh
$ llvm-readelf -S rela.fga.ro
There are 14 section headers, starting at offset 0x5e8:

Section Headers:
  [Nr] Name              Type            Address          Off    Size   ES Flg Lk Inf Al
  [ 0]                   NULL            0000000000000000 000000 000000 00      0   0  0
  [ 1] .text             PROGBITS        0000000000000000 000040 000068 00  AX  0   0 16
  [ 2] .rela.text._ZN1BD2Ev RELA         0000000000000000 000188 000048 18   I 11   1  8 <--------------------
  [ 3] .data.rel.ro      PROGBITS        0000000000000000 0000a8 000040 00  WA  0   0  8
  [ 4] .rela.data.rel.ro RELA            0000000000000000 0001d0 0000a8 18   I 11   3  8
  [ 5] .rodata           PROGBITS        0000000000000000 0000e8 000003 00   A  0   0  1
  [ 6] .comment          PROGBITS        0000000000000000 000278 000016 01  MS  0   0  1
  [ 7] .eh_frame         X86_64_UNWIND   0000000000000000 0000f0 000098 00   A  0   0  8
  [ 8] .rela.eh_frame    RELA            0000000000000000 000290 000060 18   I 11   7  8
  [ 9] .llvm_addrsig     LLVM_ADDRSIG    0000000000000000 0002f0 000005 00   E  0   0  1
  [10] .note.GNU-stack   PROGBITS        0000000000000000 0002f5 000000 00      0   0  1
  [11] .symtab           SYMTAB          0000000000000000 0002f8 0001c8 18     13   8  8
  [12] .shstrtab         STRTAB          0000000000000000 0004c0 00009d 00      0   0  1
  [13] .strtab           STRTAB          0000000000000000 00055d 000088 00      0   0  1
```

The combined RELA section gets the name `.rela.text._ZN1BD2Ev`; should it match the name of the combined section (`.rela.text`)?

This seems uncontrollable by linker script either, for example:

```
$ cat ldscript.amd64
SECTIONS {
  .text : { *(.text .stub .text.* .gnu.linkonce.t.*) }
  .rela.text : { *(.rela.text .rela.text.* .rela.gnu.linkonce.t.*) }
}
$ ./rela.sh
$ llvm-readelf -S rela.fga.ro
There are 14 section headers, starting at offset 0x5e8:

Section Headers:
  [Nr] Name              Type            Address          Off    Size   ES Flg Lk Inf Al
  [ 0]                   NULL            0000000000000000 000000 000000 00      0   0  0
  [ 1] .text             PROGBITS        0000000000000000 000040 000068 00  AX  0   0 16
  [ 2] .rela.text._ZN1BD2Ev RELA         0000000000000000 000188 000048 18   I 11   1  8
  [ 3] .data.rel.ro      PROGBITS        0000000000000000 0000a8 000040 00  WA  0   0  8
  [ 4] .rela.data.rel.ro RELA            0000000000000000 0001d0 0000a8 18   I 11   3  8
  [ 5] .rodata           PROGBITS        0000000000000000 0000e8 000003 00   A  0   0  1
  [ 6] .comment          PROGBITS        0000000000000000 000278 000016 01  MS  0   0  1
  [ 7] .eh_frame         X86_64_UNWIND   0000000000000000 0000f0 000098 00   A  0   0  8
  [ 8] .rela.eh_frame    RELA            0000000000000000 000290 000060 18   I 11   7  8
  [ 9] .llvm_addrsig     LLVM_ADDRSIG    0000000000000000 0002f0 000005 00   E  0   0  1
  [10] .note.GNU-stack   PROGBITS        0000000000000000 0002f5 000000 00      0   0  1
  [11] .symtab           SYMTAB          0000000000000000 0002f8 0001c8 18     13   8  8
  [12] .shstrtab         STRTAB          0000000000000000 0004c0 00009d 00      0   0  1
  [13] .strtab           STRTAB          0000000000000000 00055d 000088 00      0   0  1
```

File: lld/ELF/Driver.cpp (hunk 1 above)
Comment (MaskRay) [reply lld/ELF/Driver.cpp:1387]: I was thinking that `--inhibit-group-allocation` could be kept to document the default behavior. I agree that it's not useful and I will remove it.

**Comment by MaskRay** [comment]
> First, thank you so much for doing this. :)

Np:)
 
> I applied this patch to 17.0.6 to compare behavior. The group sections portion appears to work as expected, but I'm confused by one aspect of the relocation section combining:

You need commit b8dface221f4490933b0d39deb769e97ca134e5f and 0930f62cf600d9e2e9a45fef1b3a422d50be89d5 to make the `.rela.*` section name match its relocated section.

Static relocation sections copied to the output due to -r/--emit-relocs cannot be matched by input section descriptions.
`.rela.text : { *(.rela.text .rela.text.* .rela.gnu.linkonce.t.*) }` is ignored. I think GNU ld ignores them as well.

**Review by smithp35 — APPROVED** [review]
Thanks for the update LGTM.

**Comment by justincady** [comment]
> You need commit https://github.com/llvm/llvm-project/commit/b8dface221f4490933b0d39deb769e97ca134e5f and https://github.com/llvm/llvm-project/commit/0930f62cf600d9e2e9a45fef1b3a422d50be89d5 to make the .rela.* section name match its relocated section.

Thanks, that worked. I should have tested with head of line initially.

LGTM!

**Review by justincady — COMMENTED** [review]
LGTM.

**Pull request merged**